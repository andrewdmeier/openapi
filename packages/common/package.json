{
  "name": "@openapi/plugin-common",
  "version": "1.0.0",
  "license": "MIT",
  "main": "src/index",
  "dependencies": {
    "@okgrow/graphql-scalars": "^0.4.2",
    "@openapi/core": "1.0.0",
    "lodash": "^4.17.11"
  },
  "devDependencies": {
    "graphql-code-generator": "^0.15.2",
    "graphql-codegen-typescript-common": "0.15.2",
    "graphql-codegen-typescript-server": "0.15.2",
    "graphql-codegen-add": "^0.15.2"
  },
  "scripts": {
    "gen:gql": "gql-gen --config codegen.yml"
  }
}
