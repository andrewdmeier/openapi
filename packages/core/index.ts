export const utils = require('./utils');
export const openApi = require('./openApi');
export const queue = require('./queue');
export const pluginUtils = require('./pluginUtils');
export const testUtils = require('./testUtils');
