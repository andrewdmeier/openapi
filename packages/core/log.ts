import * as pino from 'pino';

const log = pino({ name: 'openapi' });

export default log;
