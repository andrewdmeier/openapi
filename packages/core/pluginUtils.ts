import * as fp from 'lodash/fp';

const log = require('./log').createLogger('@openapi/core/pluginUtils');

let pluginsByName = [];

const initPlugin = async ({ init, name }) => {
  try {
    log.info({ plugin: name }, 'initializing plugin');
    const plugin = await init();
    return {
      name,
      ...plugin,
    };
  } catch (err) {
    err.plugin = name;
    throw err;
  }
};

export const initPlugins = async pluginList => {
  pluginsByName = fp.flow(
    fp.map(fp.castArray),
    fp.map(([plugin, options = {}]) => ({
      name: plugin,
      options,
    })),
    fp.reduce(
      (acc, plugin) => ({
        ...acc,
        [plugin.name]: plugin.options,
      }),
      {}
    )
  )(pluginList);

  const plugins = Object.keys(pluginsByName).map(plugin => ({
    init: require(plugin).init,
    name: plugin,
  }));

  try {
    return Promise.all(plugins.map(initPlugin));
  } catch (err) {
    log.error(`Error while initializing plugin ${err.plugin} `, err.stack);
    throw err;
  }
};

export const getOptions = name => pluginsByName[name];
