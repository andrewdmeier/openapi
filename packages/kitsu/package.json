{
  "name": "@openapi/plugin-kitsu",
  "version": "1.0.0",
  "license": "MIT",
  "main": "src/index",
  "dependencies": {
    "@openapi/core": "1.0.0",
    "axios": "^0.18.0",
    "date-fns": "^1.29.0",
    "lodash": "^4.17.11",
    "simple-oauth2": "^2.2.1"
  },
  "devDependencies": {
    "graphql-code-generator": "^0.15.2",
    "graphql-codegen-typescript-common": "0.15.2",
    "graphql-codegen-typescript-server": "0.15.2",
    "graphql-codegen-add": "^0.15.2",
    "graphql-request": "^1.8.2"
  },
  "scripts": {
    "gen:gql": "gql-gen --config codegen.yml"
  }
}
