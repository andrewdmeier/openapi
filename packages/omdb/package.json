{
  "name": "@openapi/plugin-omdb",
  "version": "1.0.0",
  "license": "MIT",
  "dependencies": {
    "@openapi/core": "1.0.0",
    "axios": "^0.18.0",
    "dataloader": "^1.4.0",
    "lodash": "^4.17.11"
  },
  "devDependencies": {
    "graphql-code-generator": "^0.15.2",
    "graphql-codegen-typescript-common": "0.15.2",
    "graphql-codegen-typescript-server": "0.15.2",
    "graphql-codegen-add": "^0.15.2",
    "graphql-codegen-time": "^0.15.2",
    "graphql-request": "^1.8.2"
  },
  "scripts": {
    "gen:gql": "gql-gen --config codegen.yml"
  }
}
