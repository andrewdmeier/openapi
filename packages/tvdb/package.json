{
  "name": "@openapi/plugin-tvdb",
  "version": "1.0.0",
  "license": "MIT",
  "main": "src/index",
  "dependencies": {
    "@openapi/core": "1.0.0",
    "axios": "^0.18.0",
    "dataloader": "^1.4.0",
    "lodash": "^4.17.11"
  },
  "devDependencies": {
    "graphql-code-generator": "^0.15.2",
    "graphql-codegen-typescript-common": "0.15.2",
    "graphql-codegen-typescript-server": "0.15.2",
    "graphql-codegen-add": "^0.15.2",
    "graphql-request": "^1.8.2"
  },
  "scripts": {
    "gen:gql": "gql-gen --config codegen.yml"
  }
}
