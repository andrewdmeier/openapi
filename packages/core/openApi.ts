import * as fp from 'lodash/fp';
import { mergeResolvers, mergeTypes } from 'merge-graphql-schemas';

import * as pluginUtils from './pluginUtils';
import { sleep } from './utils';

const log = require('./log').default.child({ ns: '@openapi/core/openApi' });

export interface IConfig {
  db?: {
    redis?: {
      main?: any;
      queue?: any;
    };
    mongo?: any;
    postgresql: {};
  };
  queue?: {
    redisUrl: string;
  };
  enabledPlugins?: any[];
}

let openApiConfig: IConfig = {};
export const getConfig = () => openApiConfig;

export const init = async (config: IConfig) => {
  openApiConfig = config;

  const { enabledPlugins } = config;

  log.info('initializing openapi');
  try {
    // this is purely for debugging
    await sleep(process.env.NODE_ENV === 'development' ? 2000 : 0);

    log.info('initializing plugins');
    const plugins = await pluginUtils.initPlugins(enabledPlugins);
    const pluginsByName = plugins.reduce((acc, plugin) => {
      acc[plugin.name] = plugin;
      return acc;
    }, {});

    log.info('creating merged typeDefs');
    const typeDefs = mergeTypes(plugins.map(plugin => plugin.typeDefs));

    log.info('creating merged resolvers');
    const resolvers = mergeResolvers(plugins.map(plugin => plugin.resolvers));

    const context = () => (...contextArgs) => {
      const contextStore = {};

      return {
        async getContext(name) {
          if (!contextStore[name]) {
            const plugin = pluginsByName[name];
            if (!plugin) {
              throw new Error(`There is no plugin with name ${name}`);
            }

            const creator = plugin.context;
            if (!fp.isFunction(creator)) {
              throw new Error(`Plugin ${name} doesn't define any context`);
            }
            contextStore[name] = await creator(...contextArgs);
          }

          return contextStore[name];
        },
      };
    };

    return { typeDefs, resolvers, context };
  } catch (err) {
    log.error('Error while starting application: ', err.stack);
  }

  return {};
};

export default {
  init,
  getConfig,
};
