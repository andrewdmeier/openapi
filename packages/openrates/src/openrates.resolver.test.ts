import { gql, testUtils } from '@openapi/core';
import { request } from 'graphql-request';

import { name as pluginName } from '../package.json';

let server = null;

beforeAll(async () => {
  server = await testUtils.createServer([pluginName]);
});

afterAll(async () => {
  await testUtils.destroyServer(server);
});

describe('openrates', () => {
  it('should return rates for a given currency pair on a specific date', async () => {
    const query = gql`
      {
        rates(from: "USD", to: "TRY", date: "2018-11-02T00:00:00.000Z") {
          date
          from
          rates {
            to
            rate
          }
        }
      }
    `;

    const result = await request(server.endpoint, query);
    expect(result).toMatchSnapshot();
  });

  it('should return all currencies', async () => {
    const query = gql`
      {
        currencies {
          name
        }
      }
    `;

    const result = await request(server.endpoint, query);
    expect(result).toMatchSnapshot();
  });
});
