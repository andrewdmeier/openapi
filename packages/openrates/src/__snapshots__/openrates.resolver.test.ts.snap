// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openrates should return all currencies 1`] = `
Object {
  "currencies": Array [
    Object {
      "name": "HRK",
    },
    Object {
      "name": "HUF",
    },
    Object {
      "name": "IDR",
    },
    Object {
      "name": "PHP",
    },
    Object {
      "name": "TRY",
    },
    Object {
      "name": "RON",
    },
    Object {
      "name": "ISK",
    },
    Object {
      "name": "SEK",
    },
    Object {
      "name": "THB",
    },
    Object {
      "name": "PLN",
    },
    Object {
      "name": "GBP",
    },
    Object {
      "name": "CAD",
    },
    Object {
      "name": "AUD",
    },
    Object {
      "name": "MYR",
    },
    Object {
      "name": "NZD",
    },
    Object {
      "name": "CHF",
    },
    Object {
      "name": "DKK",
    },
    Object {
      "name": "SGD",
    },
    Object {
      "name": "CNY",
    },
    Object {
      "name": "BGN",
    },
    Object {
      "name": "CZK",
    },
    Object {
      "name": "BRL",
    },
    Object {
      "name": "JPY",
    },
    Object {
      "name": "KRW",
    },
    Object {
      "name": "INR",
    },
    Object {
      "name": "MXN",
    },
    Object {
      "name": "RUB",
    },
    Object {
      "name": "HKD",
    },
    Object {
      "name": "USD",
    },
    Object {
      "name": "ZAR",
    },
    Object {
      "name": "ILS",
    },
    Object {
      "name": "NOK",
    },
  ],
}
`;

exports[`openrates should return rates for a given currency pair on a specific date 1`] = `
Object {
  "rates": Object {
    "date": "2018-11-02T00:00:00.000Z",
    "from": "USD",
    "rates": Array [
      Object {
        "rate": 5.47770868,
        "to": "TRY",
      },
    ],
  },
}
`;
